src/Features/CSharp/Portable/SplitOrMergeIfStatements/CSharpIfLikeStatementGenerator.cs

```diff
@@ -0,0 +1,151 @@
+            {
+                ifLikeStatement = ifStatement;
+                return true;
+            }
+
+            ifLikeStatement = null;
+            return false;
+        }
+
+        public bool IsElseIfClause(SyntaxNode node, out SyntaxNode parentIfLikeStatement)
+        {
+            if (node is IfStatementSyntax && node.Parent is ElseClauseSyntax)
+            {
+                parentIfLikeStatement = (IfStatementSyntax)node.Parent.Parent;
+                return true;
+            }
+
+            parentIfLikeStatement = null;
+            return false;
+        }
+
+        public SyntaxNode GetCondition(SyntaxNode ifLikeStatement)
+        {
+            var ifStatement = (IfStatementSyntax)ifLikeStatement;
+            return ifStatement.Condition;
+        }
+
+        public SyntaxNode GetRootIfStatement(SyntaxNode ifLikeStatement)
+        {
+            var ifStatement = (IfStatementSyntax)ifLikeStatement;
+
+            while (ifStatement.Parent is ElseClauseSyntax elseClause)
+            {
+                ifStatement = (IfStatementSyntax)elseClause.Parent;
+            }
+
+            return ifStatement;
+        }
+
+        public ImmutableArray<SyntaxNode> GetElseLikeClauses(SyntaxNode ifLikeStatement)
+        {
+            var ifStatement = (IfStatementSyntax)ifLikeStatement;
+
+            var builder = ImmutableArray.CreateBuilder<SyntaxNode>();
+
+            while (ifStatement.Else?.Statement is IfStatementSyntax elseIfStatement)
+            {
+                builder.Add(elseIfStatement);
+                ifStatement = elseIfStatement;
+            }
+
+            if (ifStatement.Else != null)
+            {
+                builder.Add(ifStatement.Else);
+            }
+
+            return builder.ToImmutable();
+        }
+
+        public SyntaxNode WithCondition(SyntaxNode ifLikeStatement, SyntaxNode condition)
+        {
+            var ifStatement = (IfStatementSyntax)ifLikeStatement;
+            return ifStatement.WithCondition((ExpressionSyntax)condition);
+        }
+
+        public SyntaxNode WithStatement(SyntaxNode ifLikeStatement, SyntaxNode statement)
+        {
+            var ifStatement = (IfStatementSyntax)ifLikeStatement;
+            return ifStatement.WithStatement(SyntaxFactory.Block((StatementSyntax)statement));
+        }
+
+        public SyntaxNode WithStatementsOf(SyntaxNode ifLikeStatement, SyntaxNode otherIfLikeStatement)
+        {
+            var ifStatement = (IfStatementSyntax)ifLikeStatement;
+            var otherIfStatement = (IfStatementSyntax)otherIfLikeStatement;
+            return ifStatement.WithStatement(otherIfStatement.Statement);
+        }
+
+        public SyntaxNode WithElseLikeClausesOf(SyntaxNode ifStatement, SyntaxNode otherIfStatement)
+        {
+            return ((IfStatementSyntax)ifStatement).WithElse(((IfStatementSyntax)otherIfStatement).Else);
```

Comment:
a cute thing you can do is have an abstract generic base class that implements IIfLikeStatementGenerator and forwards all the methods of that to hte generic abstract versions that base classes override.  then, in the C# version for example, these are all typesafe.  not sure if that helps in VB as i imagine you can't find anything common for If and ElseIf.

Reply:
That might be cute but just like you're saying, wouldn't help for VB. I feel like all that would achieve would be moving some of the ugliness/casts into a different class. I would consider that if it affected both languages, but doing this for C# only seems a little odd.